src/map.c

```diff
@@ -3,752 +3,630 @@
-        int res = obj->comparator(new_node->key, (*indirect)->key);
-        if (res == 0) { /* If the key matches something, don't insert */
-            map_delete_node(obj, new_node);
-            return false;
-        }
-        parent = *indirect;
-        indirect = res < 0 ? &(*indirect)->left : &(*indirect)->right;
-    }
+    if (!value)
+        memset(node->val, 0, vsize);
+    else
+        memcpy(node->val, value, vsize);
 
-    *indirect = new_node;
-    rb_set_parent(new_node, parent);
-    map_fix_colors(obj, new_node);
-    map_calibrate(obj);
-    return true;
+    return node;
 }
 
-static void map_prev(map_t obj, map_iter_t *it)
+static void map_delete_node(map_t UNUSED, map_node_t *node)
 {
-    /* We have hit the end or null node. */
-    if (!it->node || it->node == obj->it_least.node) {
-        it->prev = it->node = NULL;
-        return;
-    }
-
-    if (it->node->left) { /* To the left, as far right as possible */
-        for (it->node = it->node->left; it->node->right;
-             it->node = it->node->right)
-            it->prev = it->node;
-        return;
-    }
+    free(node->key);
+    free(node->val);
+    free(node);
+}
 
-    /* If there is no left child, keep going up until there is a left child */
-    it->prev = it->node;
-    it->node = rb_parent(it->node);
+/* Constructor */
+map_t map_new(size_t s1,
+              size_t s2,
+              int (*cmp)(const map_node_t *, const map_node_t *))
+{
+    map_t tree = (map_internal_t *) malloc(sizeof(map_internal_t));
+    tree->key_size = s1;
+    tree->val_size = s2;
+    tree->cmp = cmp;
+    internal_map_new(tree);
+    return tree;
+}
 
-    while (rb_parent(it->node) && it->node->left &&
-           (it->node->left == it->prev)) {
-        it->prev = it->node;
-        it->node = rb_parent(it->node);
-    }
+/* Add function */
+bool map_insert(map_t obj, void *key, void *val)
+{
+    map_node_t *node = map_create_node(key, val, obj->key_size, obj->val_size);
+    internal_map_insert(obj, node);
+    return true;
 }
 
+/* Get functions */
 void map_find(map_t obj, map_iter_t *it, void *key)
 {
-    if (!obj->head) { /* End the search instantly if nothing is found. */
-        it->node = it->prev = NULL;
-        return;
-    }
-
-    /* Basically a repeat of insert */
-    map_node_t **indirect = &obj->head;
-
-    /* binary search */
-    while (*indirect) {
-        int res = obj->comparator(key, (*indirect)->key);
-        if (res == 0)
-            break;
-        indirect = res < 0 ? &(*indirect)->left : &(*indirect)->right;
-    }
-
-    if (!*indirect) {
-        it->node = NULL;
-        return;
-    }
-    it->node = *indirect;
-
-    /* Generate a "prev" as well */
-    map_iter_t tmp = *it;
-    map_prev(obj, &tmp);
-    it->prev = tmp.node;
+    map_node_t tmp_node;
+    tmp_node.key = key;
+    it->node = internal_map_search(obj, &tmp_node);
 }
 
 bool map_empty(map_t obj)
 {
-    return (obj->size == 0);
+    return (NULL == obj->root);
```

Comment:
Shorten as the following:
```
    return !obj->root;
```

Also, consider to mark it as "static inline" in public header.